SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DiagnoseLifetimeDependence.swift

```diff
@@ -0,0 +1,570 @@
+///
+/// In this example, the parent value is an address, so the non-escaping scope is an access scope:
+///
+///     %access = begin_access %address : $*Container
+///
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+// DiagnosticEngine already declares DiagID. Why isn't it visible?
+import ASTBridging
+public typealias DiagID = BridgedDiagID
+
+private let verbose = true
+
+private func log(_ message: @autoclosure () -> String) {
+  if verbose {
+    print("### \(message())")
+  }
+}
+
+extension InstructionRange {
+  // TODO: InstructionRange strangely does not support phis or function args. This is problematic, because the block may be mutated.
+  static func getRepresentativeInstruction(value: Value) -> Instruction {
+    if let def = value.definingInstruction {
+      return def
+    }
+    if let result = TermResult(value) {
+      return result.terminator
+    }
+    assert(Phi(value) != nil || value is FunctionArgument)
+    return value.parentBlock.instructions.first!
+  }
+}
+
+/// Diagnostic pass.
+///
+/// Find the roots of all non-escapable values in this function. All non-escapable values either depend on a NonEscapingScope, or they are produced by a LifetimeDependentInstruction that has no dependence on a parent value (@_unsafeNonEscapableResult).
+let diagnoseLifetimeDependencePass = FunctionPass(
+  name: "diagnose-lifetime-dependence")
+{ (function: Function, context: FunctionPassContext) in
+  log("Diagnosing lifetime dependence in \(function.name)")
+  
+  var diagnostic = DiagnoseLifetimeDependence(context)
+  defer { diagnostic.deinitialize() }
+  // Finding the scopes of all non-escapable values would be overkill. They are all rooted in either a non-escapable function argument or a LifetimeDependentInstruction, so use those as starting points.
+  for argument in function.arguments {
+    if !argument.type.isEscapable {
+      diagnostic.diagnose(scope: NonEscapingScope.caller(argument))
+    }
+  }
+  for instruction in function.instructions {
+    guard let dependentInst = LifetimeDependentInstruction(instruction) else { continue }
+    precondition(instruction.results.count > 0, "a lifetime-dependent instruction must have at least one result")
+    _ = dependentInst.visitDependentValues { value in
+      // use-def visitor.
+      NonEscapingScope.visitScopes(enclosing: value, context) {
+        diagnostic.diagnose(scope: $0)
+      }
+      return .continueWalk
+    }
+  }
+}
+
+/// Lifetime dependence diagnostic
+///
+/// A. For each non-escaping scope, compute its range.
+///
+/// B. Compute all values that have a lifetime dependence on the non-escaping scope and check that they fall within the scope. This performs a def-use walk which is the inverse of finding a non-escaping scope from a non-escapable value.
+struct DiagnoseLifetimeDependence {
```

Comment:
all the type and function definitions in a pass source file should be private

Reply:
Fixed. I changed everything to private. Although most of the utilities are eventually meant to be non-private. This file will have a lot of prototyping/bootstrapping until that's all figured out.